Added actor critic example

apoorvnandan opened the pull request:
A simple implementation of actor-critic method in CartPole environment.

- Solves the task in ~350 episodes. (3-4 minutes on CPU)
- Policy network implemented using Functional API. Initial layer shared between actor and critic.
- Network weights updated using `tf.GradientTape`
- Prints running reward every 10 episodes.

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+from tensorflow import math
+from tensorflow.keras import losses
+from tensorflow.keras import layers
+from tensorflow.keras import optimizers
```
You don't need these imports, just use `tf.` and `keras.` for thins like `tf.math`, `keras.losses`, etc.

apoorvnandan replied:
Got it.

I was actually wondering about the exact best practice in such cases. 
If only a single optimizer and loss function are being instantiated, we use `keras.losses.Huber` or `keras.optimizers.Adam` and if multiple things from a submodule are being used (like multiple layers), we import `from tf.keras import layers` and use `layers.Dense` etc. Is this understanding correct?

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+gamma = 0.99
+env = gym.make("CartPole-v0")
+env.seed(seed)
+eps = np.finfo(np.float32).eps.item()
```
This deserves a comment

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+    state = env.reset()
+    episode_reward = 0
+    with tf.GradientTape() as tape:  # keep track of ops for backprop
+        for timestep in range(1, 10000):
```
This should be configurable

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+while True:  # run till solved
+    state = env.reset()
+    episode_reward = 0
+    with tf.GradientTape() as tape:  # keep track of ops for backprop
```
This comment is not needed

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+            action_probs, critic_value = model(state)
+            critic_value_history.append(critic_value[0, 0])
+
+            # Sample action from action prob distribution
```
Avoid abbreviations, spell out word in full

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+            discounted_sum = r + gamma * discounted_sum
+            returns.insert(0, discounted_sum)
+
+        # Normalise
```
Normalize

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+
+        # Backprop
+        loss_value = sum(actor_losses) + sum(critic_losses)
+
```
Remove blank line (otherwise the comment seems attached to the loss sum above)

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+            diff = ret - value
+            actor_losses.append(-log_prob * diff)  # actor loss
+            critic_losses.append(
+                losses.Huber()(tf.expand_dims(value, 0), tf.expand_dims(diff, 0))
```
You don't need to reinstantiate `Huber` at every step, either instantiate it once at the start or use `keras.losses.huber` (which returns one entry per sample)

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+eps = np.finfo(np.float32).eps.item()
+
+"""
+## Implement Actor Critic network
```
Please add an intro section explaining what this script does, why it matters, and provide a link to the reference paper(s).

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+running_reward = 0
+episode_count = 0
+
+while True:  # run till solved
```
The various steps in this loop should be more abundantly commented; explain your full though process

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+
+    if running_reward > 195:  # Condition to consider the task solved
+        print("Solved at episode {}!".format(episode_count))
+        break
```
Consider adding some kind of visualization at the end, if possible

apoorvnandan replied:
I can plot the rewards for starters. Can videos be embedded in tutobooks?

fchollet replied:
Ultimately we need to be able to display the examples as a webpage. How difficult would it be to create a gif?

If easy, do it in the notebook and display the gif in-line. If difficult, then:
- Create it offline
- Include it in your PR by putting it in `site/img/examples_extra/`
- In the tutobook text, including the gif in the text via markdown tags
- Also in the tutobook text, show the code the user would normally use to display the animation

apoorvnandan replied:
Let me check about the gifs.

apoorvnandan replied:
Created the gifs. By `site/img/examples_extra/` do you mean I should create a directory called `site` in the root directory of this repo?

fchollet replied:
Here's an even simpler alternative: host the image externally, e.g. on imgur, and copy the link. Or else, just put it in `examples/rl/actor_critic_cartpole/img/` and I'll take care of hosting it appropriately.

apoorvnandan replied:
Okay.

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+"""
```
Also, we should name the folder `rl` (like `nlp`).

fchollet commented on examples/reinforcement_learning/actor_critic_cartpole.py:
```diff
@@ -0,0 +1,119 @@
+            state = tf.convert_to_tensor(state)
+            state = tf.expand_dims(state, 0)
+
+            action_probs, critic_value = model(state)
```
Also it would be more performant to put the 3 lines above in a `tf.function`

apoorvnandan replied:
Tested this. 

@tf.function takes 0.0005 seconds on an average.
The current code takes 0.0009 seconds on an average.

Both these numbers are very low compared to the overall time taken per episode, which varies from 0.01 to 0.06 seconds on an average.

fchollet replied:
Ok, we can leave it as is then.

apoorvnandan commented:
Thank you for the comments. I'll update the code.

apoorvnandan commented:
Added a bunch of annotations throughout the code. 
Added an introduction section as well. 
Please take a look.

Regarding the visualization, calling `env.render()` inside the time loop pops up a little window which shows how the agent attempts to balance the pole on the cart. I'm not sure if this can be inserted inside the tutobook? Or we can add the line and write a comment that a window will pop up with the live show?

apoorvnandan commented:
Added visualizations. Please check.

apoorvnandan commented:
Added the generated files.

fchollet commented on scripts/.ipynb_checkpoints/tutobooks-checkpoint.py:
```diff
@@ -0,0 +1,516 @@
+"""Keras tutobooks implementation.
```
Please remove this file.

apoorvnandan replied:
Ah sorry! Got added by mistake.
